### DISCUSSION TITLE
Question about partial_run

### DISCUSSION THREAD

**Discussion by igor-yusupov** [robertknight/rten#237] [Q&A] [answered]
I found interesting script for converting whisper model to onnx format: https://github.com/k2-fsa/sherpa-onnx/blob/master/scripts/whisper/export-onnx.py

In this script they convert the encoder in another way, instead of audio features, they pull the cache from the decoder: https://github.com/k2-fsa/sherpa-onnx/blob/master/scripts/whisper/export-onnx.py#L108

The main advantage is that it reduced the size of the weights by about 30%
I want to try using these weights, will your partial_run optimisation work in this case?

**Comment by robertknight** [comment, marked as answer]
The `partial_run` optimization works in any situation where there are model inputs that stay the same across each iteration of the decoder loop, and there are chunks of the graph which depend only on those inputs. For encoder-decoder transformer models this should always include the cross-attention inputs. For this model I think using `partial_run` with the `n_layer_cross_k` and `n_layer_cross_v` inputs should work, but you'll have to try it.

By the way, I recently added a new `rten-generate` crate to this repo which simplifies using transformer decoder models and applies the various key optimizations (using KV cache, `partial_run` etc.). See the `rten-examples/src/gpt2.rs` demo. It won't work for this Whisper model just yet because it hardcodes the names of some model inputs to what HuggingFace's Optimum tool uses. With a little more work to generalize the API, it should work with other models though. There are also some other improvements coming as part of this GPT-related work that should transfer to Whisper models.

**Reply by igor-yusupov** [reply]
Thank you. I'll try and come back with feedback :)

**Reply by igor-yusupov** [reply]
This implementation of whisper seems to support only recognising a single 30 second chunk without the ability to set a prompt. Probably that's why the size of scales is smaller. I'm not sure about this conclusion yet, but I haven't managed to run their test on long audio yet.

**Comment by igor-yusupov** [comment]
While parsing this repository I had 2 thoughts of how to update whisper example in https://github.com/igor-yusupov/comparisons-rten:

1) move positional embedding inside the model and take embeddings with sliсes, I think rten supports this? I also think to rewrite the structure of kv cache, and instead of 12 small tensors make 1 and select the necessary elements with sliсes inside the model.

2) I think of adding other versions of whisper, not just base. would this be useful to you?

**Reply by robertknight** [reply]
> move positional embedding inside the model and take embeddings with sliсes, I think rten supports this?

A model which takes position IDs as inputs and calculates the embeddings internally is easier to use. This is how models exported from Hugging Face's Optimum work. For the KV cache, the current structure where there are separate tensors per layer actually works well, as it means the update inside the model is simple - one `Concat` node. In the next release there will be a new optimization that takes advantage of this: https://github.com/robertknight/rten/pull/239.

> I think of adding other versions of whisper, not just base. would this be useful to you?

Yes. It would be useful to see how it scales.

**Reply by igor-yusupov** [reply]
I made a new PR: https://github.com/igor-yusupov/comparisons-rten/pull/5

Added a script to convert the model. Updated the transfer and use of `kv_cache` as I wrote above for more convenient use for different versions of models. But it turned out that this version works slower. Can you please take a look?

In general it is not a problem to use as before, that is to pass each `kv_cache` element separately, but there will be needed to do a lot of code duplication, but nevertheless of course good performance is more important

**Reply by robertknight** [reply]
Thanks for this. I should be able to take a look this weekend.

**Reply by robertknight** [reply]
Thanks for this. I had a look into the new models. The slow decoder performance is due to `ScatterND` operations  (`tensor[...] = updates` in PyTorch) that are coming from the KV cache updates. Although this could be improved in rten, this is definitely going to be slower than the single `Concat` operation (`torch.cat`) in the previous decoder model, where the relation between eg. the `k1` input and `output_k1` output looks like:

`Concat(k1, new_entry, axis=1) -> output_k1`

In addition to `Concat` being a simpler operation than `ScatterND`, and easier to optimize, the next version of rten has an optimization which allows pre-allocating these tensors with spare capacity at the start of the run, and just extending them on each decoding step without copying the whole KV-cache each time. This really only works if the KV-cache update looks like the above though.

>  but there will be needed to do a lot of code duplication

Where is the code duplication coming from? The fact that different models have different numbers of layers so you have different numbers of inputs and outputs? You could always store the KV cache details (tensor, model input, model output) in a `Vec` with one entry per layer.

The next release makes it much easier to run this kind of model though, as the rten-generate crate is able to handle the decoding loop, including managing the KV cache inputs. See https://github.com/igor-yusupov/comparisons-rten/blob/138cbe532586dbec49438bc6e9956e47d16c493e/whisper/whisper-rs/src/whisper/mod.rs#L109 for a preview, or the `gpt2.rs` example in this repository.

**Reply by igor-yusupov** [reply]
Thank you for your answer.
I updated model converting: https://github.com/igor-yusupov/comparisons-rten/commit/96334667d2d796f18d608f74f9b3ee4510539097

It works fast, but a little bit slower than the first version. Could it be because of this slice?: https://github.com/igor-yusupov/comparisons-rten/commit/96334667d2d796f18d608f74f9b3ee4510539097#diff-30d0e808e634c5be75d11e9e3f5525d1b49891c3877a22d288f6fab0624672d5L202

Thank you for idea of removing code duplication. I did it in python's converting script, so I guess I will be able to make it in Rust

UPD: I tried remove `pos_emb` from the model and sent it as before. The perfomance is the same, that doesn't seem to be the reason.

**Reply by igor-yusupov** [reply]
I updated code with vectors for `kv_cache`. But it seems that we will have to use code duplication anyway, as we need to rigidly set the number of output elements: https://github.com/igor-yusupov/comparisons-rten/blob/update_converting_whisper/whisper/whisper-rs/src/whisper/mod.rs#L310

But I understand this is one of the features of Rust, it doesn't really like polymorphism, but the speed of execution is much higher, isn't it?

**Reply by robertknight** [reply]
Use [`Model::run`](https://docs.rs/rten/latest/rten/struct.Model.html#method.run) instead of `run_n` if the number of output elements is dynamic. The reason for having `run_n` is just convenience in the case when the number of outputs is statically known, as you can do `let [output_1, output_2] = model.run_n(...)`.

For high level APIs like `Model::run`, using fixed-length arrays vs dynamic-length slices doesn't affect performance. In very hot loops on the other hand (eg. in the implementation of operations like adding tensors), using fixed-sized arrays can enable the compiler to generate much more efficient code.

**Reply by igor-yusupov** [reply]
Thank you. I did it and now I'm ready to add other versions of Whisper

**Reply by igor-yusupov** [reply]
I have added ability to run 2 versions of whisper: Tiny and Small. I didn't check how it runs with Medium and Large models cuz they are too big. But I can add them if you want.
I hope it helps you in developing your engine. It already works awesome with Whisper model.

My new code runs a little slower than the old version. About 0.07 - 0.1 second slower on my machine. Haven't figured out yet what this could be, it is possible that the Rust compiler is a bit worse at optimising for vectors than for variables. I would be grateful if you could look at the code and tell me what could be improved :)

**Reply by robertknight** [reply]
Sorry for the late response. Thanks for this. I'll take a look a bit later in the week.

>  I didn't check how it runs with Medium and Large models cuz they are too big

Support for models larger than 2GB will be coming in the next release. See https://github.com/robertknight/rten/pull/260. For decoder models which are constrained by memory bandwidth, performance decreases roughly linearly as the file size gets larger (eg. GPT-2 XL is 4x the size of GPT-2 medium and runs at about 1/4 the speed).

**Reply by robertknight** [reply]
> My new code runs a little slower than the old version. About 0.07 - 0.1 second slower on my machine. Haven't figured out yet what this could be

I had a look at the new model. It looks like there is an easy change that can be made to the model to improve performance.

In the `MultiHeadAttention` block there is a `scale` variable which is calculated from the `q` shape and used to scale the q, k and v tensors. In cross attention, the `k` and `v` tensors come from the encoder output and the `q` from the decoder. Ideally we want to precompute as much of the decoder graph as possible from just the encoder outputs. Since the `k` and `v` tensors don't change at each iteration, we should be able to compute the `k = k.view(...) * scale` and `v = v.view(...) * scale` terms once and re-use them at each step. However since `scale` is computed from the shape of `q` it could in theory change at each iteration. It doesn't, but because it _could_, RTen has to compute the scaled `k * scale` and `v * scale` on every decoder step.

If we change the model so that the `scale` value is computed only from constants, then the `k * scale` and `v * scale` terms will be computed just once and re-used on each decoder step:

```diff
diff --git a/whisper/src/decoder.py b/whisper/src/decoder.py
index 51fd6a2..7cce6f7 100644
--- a/whisper/src/decoder.py
+++ b/whisper/src/decoder.py
@@ -43,10 +43,12 @@ class MultiHeadAttentionWithCache(nn.Module):
         self.value = nn.Linear(n_state, n_state)
         self.out = nn.Linear(n_state, n_state)
         self.n_ctx = n_ctx
+        self.n_state = n_state
 
     def qkv_attention(self, q: Tensor, k: Tensor, v: Tensor):
-        _, n_ctx, n_state = q.shape
-        scale = (n_state // self.n_head) ** -0.25
+        _, n_ctx, _ = q.shape
+
+        scale = (self.n_state // self.n_head) ** -0.25
         q = q.view(*q.shape[:2], self.n_head, -1).permute(0, 2, 1, 3) * scale
         k = k.view(*k.shape[:2], self.n_head, -1).permute(0, 2, 3, 1) * scale
         v = v.view(*v.shape[:2], self.n_head, -1).permute(0, 2, 1, 3)
@@ -89,6 +91,7 @@ class MultiHeadAttention(nn.Module):
         self.key = nn.Linear(n_state, n_state, bias=False)
         self.value = nn.Linear(n_state, n_state)
         self.out = nn.Linear(n_state, n_state)
+        self.n_state = n_state
 
     def forward(
         self,
@@ -107,8 +110,8 @@ class MultiHeadAttention(nn.Module):
     def qkv_attention(
         self, q: Tensor, k: Tensor, v: Tensor, mask: Optional[Tensor] = None
     ):
-        _, n_ctx, n_state = q.shape
-        scale = (n_state // self.n_head) ** -0.25
+        _, n_ctx, _ = q.shape
+        scale = (self.n_state // self.n_head) ** -0.25
         q = q.view(*q.shape[:2], self.n_head, -1).permute(0, 2, 1, 3) * scale
         k = k.view(*k.shape[:2], self.n_head, -1).permute(0, 2, 3, 1) * scale
         v = v.view(*v.shape[:2], self.n_head, -1).permute(0, 2, 1, 3)
```

## Profiling notes

Some notes about how I figured this out:

When enabling timing metrics by setting the `RTEN_TIMING` environment variable to 1, the output shows that quite a bit of time in the decoder is being spent in `Mul` operations:

```
Graph run of 1498 ops finished in 19.161ms
Pool allocs 888 hits 817
MatMul                 9.65ms (50.36%)
Mul                    4.94ms (25.79%)
FusedTranspose(MatMul) 1.53ms (8.00%)
```

Setting `RTEN_TIMING=by-shape=1` shows more detailed information about the input shapes. It also contains the name of one of the graph nodes in that group.


```
Mul                    7.01ms (32.55%)

    Shape                       Count  Mean (ms)  Total (ms)  ns/input elem  Example node
    --------------------------  -----  ---------  ----------  -------------  ----------------------------------
    [], [1, 8, 64, 1500]        6      1.146      6.874       1.492          /decoder/blocks.0/cross_attn/Mul_1
    [2], []                     36     0.001      0.035       324.565        /decoder/blocks.0/attn/Mul_4
    [1, 8, 64, 86], []          6      0.006      0.035       0.132          /decoder/blocks.0/attn/Mul_1
```

The node names here correspond to the `nn.Module` structure in PyTorch, so `/decoder/blocks.0/cross_attn/Mul_1` is a multiplication operation inside an `nn.Module` named `cross_attn`. You can see where the node is inside the graph using [Netron](https://netron.app). From there you need to try and locate the coresponding code in the model.

There are some [more notes on profiling here](https://github.com/robertknight/rten/blob/main/docs/performance.md).

**Reply by robertknight** [reply]
After making the above change and re-exporting + converting the model, we can see that the amount of time spent in each decoder step in `Mul` operations drops:

```
Graph run of 1408 ops finished in 12.614ms
Pool allocs 826 hits 781
MatMul                 8.85ms (70.19%)
FusedTranspose(MatMul) 1.09ms (8.63%)
Transpose              0.55ms (4.34%)
Reshape                0.49ms (3.91%)
Concat                 0.36ms (2.84%)
```

It is now spending ~80% of the time in MatMul operations, which is a healthy place to be. The `FusedTranspose(MatMul)` output is part of a new optimization in the next release, so you'll see slightly different output in the stable version.

**Reply by igor-yusupov** [reply]
Thank you so much!
I updated the code according to your edits and it started working even faster than the previous one. Then I saw on changelog that some operators have been optimized and decided to update the rten version to 0.11.0 and the inference became even faster. I think now it's faster than whisper.cpp, isn't it? It's hard to imagine how much faster it would be if it were possible to run quantized models.

Thanks for sharing how you used the profiler to find bottlenecks. It is a powerful tool in your engine.

I'll try to add the inference capability of other versions of whisper then, and also look at the code from rten-generate to use optimizations from there for inference.

**Reply by robertknight** [reply]
> I think now it's faster than whisper.cpp, isn't it?

On my laptop CPU (2020 4-core Intel i5) the two are very close, although there is quite a big difference in where the time is spent. The decoder is faster in whisper.cpp but the encoder is slower, so the two balance out. There are still some more opportunities for fusion (combining operators together), but indeed I think quantization support is needed to make a big improvement for the larger models.

**Reply by igor-yusupov** [reply]
I have finished adding the remaining versions of Whisper (the only one not added is the Large version; it converts without any issues but, for some reason, it doesn't work correctly with greedy search).

I have also added a script for quantization, so now it's easy to quantize models and run inference with them in Python. Interestingly, quantized models run slower on my machine. I used dynamic quantization, as recommended for transformers in the official ONNX Runtime [documentation](https://onnxruntime.ai/docs/performance/model-optimizations/quantization.html#method-selection) but it might be worth trying static quantization, although it requires generating data for it.

**Reply by robertknight** [reply]
Cool, thanks for that! What problems did you encounter with the large model? Was the output junk or repeating itself or something else?

Regarding [the FAQ comment on performance](https://onnxruntime.ai/docs/performance/model-optimizations/quantization.html#why-am-i-not-seeing-performance-improvements) my x64 CPU does have VNNI (aka. "Deep Learning Boost") and I see the quantized model (base size) in your repository being about 2x as fast as the non-quantized model. This is expected as VNNI replaces a sequence of three instructions used in int8 dot products with one. On older CPUs I'd expect performance might be better or worse depending on whether inference is compute bound or memory bound. LLMs for example are usually memory bound during inference, so quantization is always helpful there.

**Reply by igor-yusupov** [reply]
Yeah, the output repeating itself on the second chunk of audio without getting EOS token

What about ARM processors? Can the quantized model run slower on them? It runs about 15% slower on my machine. I see that the FAQ section says it should be faster on ARM, but for some reason, I'm getting the opposite result

**Reply by igor-yusupov** [reply]
Nevertheless, sacrificing 15% speed for a 4-fold reduction in model size is quite a good deal :)

**Reply by robertknight** [reply]
I'm not as familiar with Arm, but my understanding is that performance depends on the availability of dot product instructions (SDOT, UDOT). These have been available on chips released since ~2018 (including Apple M-series, Arm Cortex A-76 and later).

**Reply by robertknight** [reply]
> Nevertheless, sacrificing 15% speed for a 4-fold reduction in model size is quite a good deal :)

Yes. The memory size reduction is a big plus, especially once you get into models with 500M+ params.

**Reply by igor-yusupov** [reply]
> These have been available on chips released since ~2018 (including Apple M-series ...

I use Apple's M-series processor and I even got curious about why the result is like this. It's possible that on these processors, working with fp32 is also highly optimized, while as far as I understand, quantized models perform more operations. I think there is also a possibility that during the conversion, an architecture is created that is more optimized for x86-64, or that the Apple Neural Engine is utilized more when working with fp32.

In general, I'll try to figure it out, and if I find something interesting, I'll come back with feedback.
